lib/std/heap.zig

```diff
@@ -8,6 +8,129 @@ const c = std.c;
+    // Source: Wikipedia "Page (computer memory)"
+    .powerpc, .powerpc64, .powerpc64le, .powerpcle => 64 << 10,
+    .riscv32, .riscv64 => 4 << 10,
+    .sparc, .sparcel => 256 << 10,
+    .sparc64 => 64 << 10,
+    // Rare architectures with little support.
+    else => @compileError("Does pageSize() apply to this architecture? If so, file an issue."),
+};
+
+/// Compile time known minimum page size.
+pub const page_size: usize = page_size_os orelse switch (builtin.cpu.arch) {
+    // Common knowledge.
+    .wasm32, .wasm64 => 64 << 10,
+    .x86, .x86_64 => 4 << 10,
+    .thumb, .thumbeb, .arm, .armeb, .aarch64, .aarch64_32, .aarch64_be => 4 << 10,
+    // Explicitly only 4kb.
+    // https://refspecs.linuxbase.org/ELF/zSeries/lzsabi0_zSeries.html#AEN798
+    .s390x => 4 << 10,
+    // Source: Wikipedia "MIPS" and Linux mips/Kconfig.
+    .mips, .mipsel, .mips64, .mips64el, .loongarch32, .loongarch64 => 4 << 10,
+    // Source: csky/Kconfig only selects HAVE_PAGE_SIZE_4KB.
+    .csky => 4 << 10,
+    // Hexagon's Kconfig/page.h in Linux. Non-huge pages go past 256KB.
+    .hexagon => 4 << 10,
+    // Source: Zig's own libc page.h for arc.
+    .arc => 4 << 10,
+    // Source: Wikipedia "Page (computer memory)"
+    .powerpc, .powerpc64, .powerpc64le, .powerpcle => 4 << 10,
+    .riscv32, .riscv64 => 4 << 10,
+    .sparc, .sparcel => 4 << 10,
+    .sparc64 => 8 << 10,
+    // Rare architectures with little support.
+    else => @compileError("Does pageSize() apply to this architecture? If so, file an issue."),
+};
+
+var runtimePageSize = std.atomic.Value(usize).init(0);
+
+/// Runtime detected page size.
+pub inline fn pageSize() usize {
+    if (@inComptime()) {
+        @compileError("pageSize() must NOT be used in comptime. Use page_size variants instead.");
+    }
+    if (page_size == page_size_cap) {
+        if (queryPageSize() != 0)
+            assert(queryPageSize() == page_size);
+        return page_size;
+    }
+    const size = queryPageSize();
+    std.debug.assert(size > 0);
+    return size;
+}
+
+// Runtime queried page size.
+fn queryPageSize() usize {
+    var size = runtimePageSize.load(.unordered);
+    if (size > 0) return size;
+    defer {
+        if (size != 0) {
+            std.debug.assert(size >= page_size);
+            std.debug.assert(size <= page_size_cap);
+        }
+        runtimePageSize.store(size, .unordered);
+    }
+    switch (builtin.os.tag) {
+        .linux => size = if (builtin.link_libc) @intCast(std.c.sysconf(std.c._SC.PAGESIZE)) else std.os.linux.getauxval(std.elf.AT_PAGESZ),
+        .macos => blk: {
+            size = @import("c/darwin.zig").machTaskForSelf().getPageSize() catch break :blk;
+        },
+        .windows => {
+            var info: std.os.windows.SYSTEM_INFO = undefined;
+            std.os.windows.kernel32.GetSystemInfo(&info);
+            size = info.dwPageSize;
+        },
+        else => if (@hasDecl(std.c, "_SC") and @hasDecl(std.c._SC, "PAGE_SIZE")) {
```

Comment:
`@hasDecl()` is usually a code smell now; see #20679 for the way it should be done going forward.